### ISSUE TITLE
Shadow lieutenant will bugged kill you

### ISSUE THREAD

**Issue by KHeket** [CleverRaven/Cataclysm-DDA#68224]
### Describe the bug

If you became a target of shadow lieutenant and also coming to Exodii castle - you will be teleported several times and killed because of falling damage

### Attach save file

[1-trimmed.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/12667489/1-trimmed.tar.gz)
Save file for testing. I teleported character to the Rubik castle and change to night, all you need is summon shadow lieutenant

### Steps to reproduce

1. Stay not far from rubik castle
2. Wait/change time to night
3. Wait for shadow lieutenant coming or summon it by debug menu
4. Go to Rubick castle
5. See random teleports
6. Die because of fall damage

### Expected behavior

Not buggy die

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19045.3448 (22H2)
- Game Version: 9506294 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth]
]

**Comment by KHeket** [comment]
@RenechCDDA can you check pls? And I also noticed, that if I kill Rubick and Liliya through debug menu - no bug appeared
I also tested the same at hub 01 - the same bug appeared. Summoned shadow lieutenant near the hub, walk closer to hub - random teleportation and death because of fall damage
With Refuge center I cannot make the same

**Comment by SariusSkelrets** [comment]
Can confirm the shadow lieutenant can reappear for a moment and chain teleport you, spamming the "you can't outrun your own shadow" message

Got it once at the Exodii castle and another time on the black glass at the bottom of that lab

**Comment by Cupidatis** [comment]
Like this but in evac shelter near a lab.

**Comment by RenechCDDA** [comment]
I'm not able to reproduce this exactly, but I have a guess as to how this is happening. (I am instead getting "invalid beta talker" everytime it pops up, which is a problem along the same vein.)

It sounds like the alpha and beta talkers for the teleport EOC are being swapped. From the suggestions that it happens near the exodii (and an evac shelter), I'm guessing it's somehow running on NPCs of different factions.

 No idea for the black glass instance though.

**Comment by RenechCDDA** [comment]
Oooooh okay, the teleporting (OF YOU) happens on the non-msvc builds. This just got very ugly.

**Comment by huix11** [comment]
r

**Comment by RenechCDDA** [comment]
@lispcoc Can you figure out what's going on here? I believe that EOC_LIEUTENANT_SHADOW_CATCHUP is activating for NPCs (it should not be!) or `Shadow_Catchup_Pos`'s value is being incorrectly set by NPCs (it should not be!), but if so it seems to be an error on the C++ side? Neither of those has `global` defined (defaults false), and neither do they have `run_for_npcs` (defaults false)... and yet the presence of NPCs causes incorrect values to be set.

(Of note: It seems Rubik is also being teleported onto the same tile as the player, which is definitely not right!)

**Comment by lispcoc** [comment]
u_location_variable returns the player's own location if the target monster cannot pass through. I think this makes the player himself a beta talker in run_eoc_with.

You can avoid teleporting the player by executing run_eoc_with only when u_location_variable is successful.

```
  {
    "type": "effect_on_condition",
    "id": "EOC_LIEUTENANT_SHADOW_CHECK_POS",
    "recurrence": 1,
    "condition": { "math": [ "u_monsters_nearby('mon_lieutenant_shadow', 'radius': 60)", ">", "0" ] },
    "effect": [
      {
        "u_location_variable": { "global_val": "Shadow_monster" },
        "monster": "mon_lieutenant_shadow",
        "target_min_radius": 0,
        "target_max_radius": 60,
        "true_eocs": [
          {
            "id": "EOC_LIEUTENANT_SHADOW_CHECK_POS_OK",
            "effect": { "run_eoc_with": "EOC_LIEUTENANT_SHADOW_CATCHUP", "beta_loc": { "global_val": "Shadow_monster" } }
          }
        ]
      }
    ]
  },
```

**Comment by lispcoc** [comment]
> @lispcoc Can you figure out what's going on here? I believe that EOC_LIEUTENANT_SHADOW_CATCHUP is activating for NPCs (it should not be!) or `Shadow_Catchup_Pos`'s value is being incorrectly set by NPCs (it should not be!), but if so it seems to be an error on the C++ side? Neither of those has `global` defined (defaults false), and neither do they have `run_for_npcs` (defaults false)... and yet the presence of NPCs causes incorrect values to be set.


I looked into the exact specifications of the "global" EOC and they are as follows. This is difficult to read from the documentation and may be counterintuitive to the word "global"

| global | Runs for                | Activate/Deactivate state |
|---------|--------------------|-----------------------------|
| true      | avater only          | depend on avater              |
| false     | avater and NPCs | depend on each character |

In other words, if you want to enable "EOC_LIEUTENANT_SHADOW_CHECK_POS" only for avatar, I think you need to set global to true.
